Initial import: Testing framework for Unikraft builds (GSoC 2025)

shank250 opened the pull request:
This PR imports the testing framework developed during GSoC 2025 for automating and validating Unikraft builds.

shank250 committed:
initial: Migrate codebase from unikraft-upb/catalog

Migrate the codebase from razvand/generator/new-design
branch of unikraft-upb/catalog repository into a standalone project
The initial work by Razvand Deaconescu can be found at https://github.com/unikraft-upb/catalog/tree/razvand/generator/new-design/utils/new-design

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
chore: Add configuration files for linters

Add default configrations for linters and update .gitignore
with virtual environment

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
chore: Update requirements.txt with latest dependencies

Add all the python dependent library with their
version.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
style: Reformat codebase using black and organize imports with isort

Restructure the codebase to consistent style with black and
isort.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
docs: Add missing module-level docstrings to resolve pylint C0114

Add docstrings to each of the python files.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
refactor: Restructure codebase with pylint

Based on the pylints wanrning this commit updates the codebase to follow
PEP 8 standard.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
refactor: Add static method and improvise return statement

Resolve pylint warning regarding ussage of static method and initialise variable before if condition to better handle return value.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
refactor: Introduce support for .app fixture folder

Introduce a new .app fixture folder to store test application-related files and update .gitignore to include these new files.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
chore: Fix Pylint warning for singleton-comparison

Using `is` and `is not` for singleton comparison.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
chore: Add /doc/current_design.txt

Contains overview of the general design of the testing framework.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
chore: Add setup.sh and cleanup.sh for simpler dev/testing

Add `setup.sh` and `cleanup.sh` for simpler development and testing. These files provide quick cleanup of resources and program initialization.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Implement port number extranction from README.md

Extract port number from .app/README.md and populate the value to config.yaml.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Improvise cleanup, setup and test_generation for devlopment

Add a main runner file for quick application testing to aid in developing the testing framework.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
refactor: Reformat codebase with black & isort

Update codebase to improvise the code style and import statmment order according to
predefined standard linters.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Implement TestRunner and add loggers

Introduce a basic implementation of the `TestRunner` class to handle build and run steps for targets with logging.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Implements _run_target() method

THe specific target running logic is implemented and linters are used to
format the code.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
docs: Update the readme.md to add Sudo Setup

Add instrauction on how to skip password prompt on ussage of specific
sudo commands.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
revert: Remove loggers from the script

Removes direct ussage of loggers in each of the file.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Implement _test_target_run() and overall runs of kernel

Implement the logic for running unikernel with networking enabled.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
typo: Update config.yaml

Fix the typo mistake in the config file of networking

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
typo: Update src/test_runner.py

Fix spelling of started in logging.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update src/test_runner.py for proper return type

Improve exception handling and returns the file_path

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Implement creation of rootfs from app_fs_init.sh

Creates initfs.cpio file from app_fs_init.sh & copy it to the framework directory.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Implement exception handling for initrd.cpio file not found

Handles error  if framework fails to create initrd.cpio file from
app_fs_init.sh

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update the successful check of unikernel build

Update the implementation to verify a successful unikernel build by checking the path to the build file.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update testing application path

Update the code to pass the correct path of the application directory
when generating app_fs_init.sh file.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Impletement extensive .app/README.md parser

Provides a process of extracting meta data related to the
testing application like memmory, port no, arch etc. Update the
README.md with latest information.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update the applications path in Kraftfile

Rootfs path doesn't contains .app/ directory in the kraftfile generated for make build with kraft run.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update parser logic to better extract memory info

Remove 'M' from the parsed memory value populated in config.yaml.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
config: Update .rootfs-for-testing in .gitignore

Add path to the .rootfs-for-testing in .gitignore file.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
docs: Update readme.md with setup and ussage

Add content on basic Setup and Ussage of testing framework

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
refactor: Update support for app dir from relative to absolute

Update codebase to require an absolute path for the application directory instead of a relative path. This simplifies application path identification during testing.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
docs: Add CONTRIBUTING.md with guidelines for contributing to the testing framework

Add basic guidelines for contributing to the testing framework.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
docs: Add COPYING.md with BSD 3-Clause License

This file is taken from unikraft repository

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
refactor: Add loggging statements for better understanding

Add loggging statemented to better understand the flow and statte of
testing framework

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
docs: Update README.md with clear insctructions

Update include overview, setup information, contributions, suggestions
etc.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Delete CONTRIBUTING.md

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Add test application dir replica i.e. test-app-config

Introduce a replica of the catalog repository inside the `test-app-config` directory to properly manage configuration files for individual applications.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Add llm support for RunConfig.yaml generation

Add functionality to generate a RunConfig.yaml file using an LLM. The LLM processes the application's README.md file and its details to create the configuration file from the response.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Support for loading env variable relateed to LLM API

Add support for loading API keys from environment variables (currently supports Groq and OpenAI).

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Implement logic to log the build and run

Implement logic to log the build and run of every target and generate an overall build and run report.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Add run with curl command test and validation

Update unikernel run test step in 3 catagories curl command, list of
commands or no commands. This commit specifically add support for curl
command test and validation.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update logging inplace of print statements

Ussage of logging by extending the Logger class inplave of print
statements.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update with isort, black and remove _test_target_run()

Update code sytle with linters like isort, black and remove not required
code.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Support for run test of `no-command` and `list-of-command`

The application having no command and list of command for runs have got support.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update curl run test for tap network same as bridge

Provide same testing functionality in both networking(tap & bridge)

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Add sudo privilage to remove .tests and .app

Add sudo for deleting fixture folders like .app & .tests

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Implement setup of .sessions

The .sessions directory are used to store run & build logs for each of
the testing session.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update rootfs path for kraft builds

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Split the tester_config.yaml

Based on the suggestion of razvand tester_config.yaml is split into
variants.yaml and config.yaml for better modularity.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update requirements.txt for LLM  support

Add required Python libraries for LLM support used to create RunConfig.yaml for new applications.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Add qemu-system-aarch64 in sudo setup

Update README.md in sudo setuo instructions.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Add -t parameter for testing specific targets

Support for testing specific target is with its no is given.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Add Build and Run Config files for catalog/library apps

Initial setup config files for catalog/library are added

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update run networking configration

Improves code  for nat & bridge run command update in `list-of-commands`

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update redis config file

Increate build time & add more expected output

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update README.md & application path

README.md is updated with CLI instructions and path to the .app is
updated.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

razvand committed:
scripts: Use the same path for linker and compiler

Use the `LD` variable to define the linker path when using `make`.
Use the same value as the `CC` variable (used to define the compiler).
This is required to use the same linker and compiler version. If the
version differs, linker errors may appear.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

shank250 committed:
chore: Add library and examples config files

Adds RunConfig and BuildConFig file for library and example
applications.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Remove library in kraftfile for examples

The kraftfile generated for examples contains library, which is not
required. So, this commit removes those libraries from kraftfile.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
chore: Update logs to .gitignore

Adds .sessions and log files to .gitignore

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Add configurable tests directory support with --tests-dir parameter

Provides support for generating custom test directory names. Inplace of
.tests we can provide any other name in the CLI

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Add configurable app directory support with --app-dir-name parameter

Support for initialising framework with custom name for application
directory. Inplace of .app we can pass any other custom name in CLI

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Support for Early exit (-g flag) is implemented

Generate-only (-g) mode enables users to only generate test files and
the framework would not build and test any application automatically.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Implement runtime kernel target generation

Generation of only runtime kernel helps in examples supports.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Add runtime kernel generation and management functionality

Provids support for generating and managing the runtime kernels for
catalog/examples.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Update run scripts for vmm based builds & complete pipeline for vmm tests

Run scripts are updated to properly boot the rootfs for library and
examples differently.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fiX: Kill buildkit process after kraft build

Sometimes kraft fails to kill the buildkitd process. So, we need to kill
after each ussage of kraft commands.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Support for all runtime kernels other than base

Initial runtime kernel setup was only able to create base runtime
kernel. Now with this commit we can create any runtime kernel which is
present in the catalog/library.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Implement kernel pull from local registry for kraft build and vmm run

Implement the process of building a runtime kernel with `kraft`, pushing it to a local registry, and then pulling it with `kraft` to be stored in the `runtime_kernel` folder. This process is necessary because direct build and usage as a runtime is not supported in `kraft`.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
docs: Update README.md with Local Docker Registry Setup

Provid general instructions on how to step local registry for testing
with catalog/examples

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Add Makefile.uk if not present & check buildkitd process

Add a simple Makefile.uk if one does not exist to handle cases where applications in the catalog are missing the file.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
hotfix: Update logic for selecting kernel from .unikraft/build/... with make build

Update the logic for selecting the built kernel from the `.unikraft` directory and copying it to the `runtime_kernel` directory. This feature allows example runtimes to be created and used from a central location.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Support for all runtime kernels for catalog/examples

Handle the creation and storage of runtime kernels for catalog/examples.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Initial github workflow setup

Receive dispatched data (PR metadata) from the source repository.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update the logic for find_kernel_file

Fixes the selection of kernel from .unikraft/builds/ folder by searching
x86_64 in the kernel name. It may fail for aarch64 architecture.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Setup of repository with dependency in Workflow

Add steps for installing required dependencies and general checks.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Add catalog and kraftkit setup in Workflow

Add support for cloning the catalog repository and setting up the kraft CLI tool (for Github Workflow).

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
chore: Add catalog/examples application config file

Add RunConfig.yaml and BuildConfig.yaml file pair.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
feat: Add overall_test.py for testing all the applications at once

Wrapper script designed to be used by workflows for automatically
running tests on all the library applications.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

razvand committed:
README: Improve formatting

Do the following actions:
- Use only one sentence per line.
- Remove unnecessary bolded words.
- Use proper code block types (`console`, `text`).
- Properly indent list contents.
- Use backtick formatting (typewritter format) where required.
- Remove trailing white spaces.
- Use `.venv/` as the Python virtual environment directory.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
config: Present a template configuration file instead of a complete one

Replace `config.yaml` with `config.yaml` and ignore `config.yaml`. This
is to instruct the user to create and update their own configuration
file. Because the file is ignore, it will not appear as a Git-followed file.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
README: Introduce Requirements section

Introduce section with package and configuration requirements for the
testing framework.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

shank250 committed:
fix: Simplify test result status logic in run_test method

Remove partial-pass status from the test result logic in the run_test method.

Signed-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
remove: Delete main.sh file

Main.sh file was designed for fast testing execution while development.
Now the framework is having more CLI options with src/main.py. As such,
remove main.sh file.

Sign-off-by: Shashank Srivastava <shashank21005@gmail.com>

razvand reviewed (changes requested):
First line of commit messages looks good.

For the commit description paragraph:
- Use direct voice: Instead of "This commit migrates" say "Migrate".
- Use imperative. Instead of "Adds path" use "Add path".
- Do not use 1st person, use imperative 2nd person. Instead of "Thus, we remove main.sh file" use "As such, remove main.sh file.".
- Do not use passive voice, use active imperative voice at 2nd person. Instead of "Set of RunConfig.yaml and BuildConfig.yaml files are added." use "Add RunConfig.yaml and BuildConfig.yaml file pair."
- Use / Start with verbs. "Command" the commit do do something. E.g. instead of "Support for initialising framework with custom name for application directory" use "Introduce support for ..."
- When introducing something new, use the verb "Introduce". When adding something to an existing component, use the verb "Add". When modifying / improving an existing component, use the verb "Update" or "Improve" or "Change". When fixing an issue, use the verb "Fix".

Very good work with the semantic commit prefixes ("fix:", "chore:", "fix:", "hotfix:", "docs:".

Generally, be consistent in commit messages.

razvand commented:
When testing [`nginx/1.25`]() with bridged networking this happens:

```console
$ cat tests/00001/run-01/complete_run.log
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (7) Failed to connect to 172.44.0.2 port 8080 after 1 ms: Could not connect to server

[✗] Curl command failed
```

This is when using bridged networking:

```console
$ cat tests/00001/run-01/config.yaml 
hypervisor: kvm
networking: bridge
rootfs: none
run_tool: vmm

$ cat tests/00001/config.yaml 
base: /home/monkey/razvand/orgs/unikraft/maintainer-tools/workdir
arch: x86_64
bootloader: multiboot
build_tool: make
compiler:
  path: /usr/bin/gcc-14
  type: gcc
debug: none
pie: false
platform: fc
vmm: /usr/local/bin/firecracker-x86_64
```

This is because the `curl` test uses port `8080` (i.e. the `PublicPort`) instead of port `80` (i.e. the `ExposedPort`). For bridged networking, the `ExposedPort` should be used; for NAT neworking, the `PublicPort` must be used:

```console
$ cat test-app-config/catalog/library/nginx/1.25/RunConfig.yaml 
RunMetadata:
  Memory: 256
  Networking: true
  ExposedPort: 80
  PublicPort: 8080

TestingType: curl
ListOfCommands:
  - curl localhost:8080

ExpectedOutput:
  - "Welcome to nginx!"
  - "Hello World!"
  - "Powered by"
  ```

razvand commented:
When running the `build` script, there may be issues with removing a given directory. This is because a previous phase may have been run by the `root`:

```console
monkey@dev-fieldeng0:~/.../autokraft/.tests/00001$ cat build 
#!/bin/sh

test -d /home/monkey/razvand/orgs/unikraft/autokraft/.tests/00001/.unikraft/build || mkdir -p /home/monkey/razvand/orgs/unikraft/autokraft/.tests/00001/.unikraft/build
make -f /home/monkey/razvand/orgs/unikraft/autokraft/.tests/00001/Makefile distclean
UK_DEFCONFIG=/home/monkey/razvand/orgs/unikraft/autokraft/.tests/00001/defconfig make -f /home/monkey/razvand/orgs/unikraft/autokraft/.tests/00001/Makefile defconfig
touch Makefile.uk
make -f /home/monkey/razvand/orgs/unikraft/autokraft/.tests/00001/Makefile prepare
make CC=/usr/bin/gcc-14 LD=/usr/bin/gcc-14 -f /home/monkey/razvand/orgs/unikraft/autokraft/.tests/00001/Makefile -j $(nproc)
monkey@dev-fieldeng0:~/.../autokraft/.tests/00001$ ./build
make[1]: Entering directory '/home/monkey/razvand/orgs/unikraft/maintainer-tools/workdir/unikraft'
  RM      build
rm: cannot remove '/home/monkey/razvand/orgs/unikraft/autokraft/.tests/00001/.unikraft/build/rootfs-cache/ingest': Permission denied
rm: cannot remove '/home/monkey/razvand/orgs/unikraft/autokraft/.tests/00001/.unikraft/build/rootfs-cache/blobs/sha256/09e235bc8e540977558c701fb8ab59fa4200261cbdb1084873683643b20bc72a': Perm
ission denied
rm: cannot remove '/home/monkey/razvand/orgs/unikraft/autokraft/.tests/00001/.unikraft/build/rootfs-cache/blobs/sha256/09fb74cf08999d831428815830f2e8335f00ef38455d3df459e549f4642d8ffb': Perm
ission denied
[...]
```

The solution is to prefix all remove / clean commands with `sudo`.

razvand commented:
The `README.md` file will be have to be updated to feature the new URL of the repository (https://github.com/unikraft/autokraft) instead of the old one (https://github.com/shank250/testing-framework-uk-build.git).

razvand commented:
I couldn't run an example:

```console
$ python src/main.py ~/razvand/orgs/unikraft/catalog/examples/httpserver-nodejs21
2025-09-20 21:54:56,632 - INFO - Main Started
2025-09-20 21:54:56,632 - INFO - Main Started
2025-09-20 21:54:56,632 - INFO - Running cleanup script
2025-09-20 21:54:56,632 - INFO - Running cleanup script
2025-09-20 21:54:59,202 - INFO - Running setup script for /home/monkey/razvand/orgs/unikraft/catalog/examples/httpserver-nodejs21
2025-09-20 21:54:59,202 - INFO - Running setup script for /home/monkey/razvand/orgs/unikraft/catalog/examples/httpserver-nodejs21
['/home/monkey/razvand/orgs/unikraft/autokraft/scripts/utils/setup.sh', '/home/monkey/razvand/orgs/unikraft/catalog/examples/httpserver-nodejs21']
2025-09-20 21:54:59,600 - INFO - Generated session name: session_20_09_2025_21_54
2025-09-20 21:54:59,600 - INFO - Generated session name: session_20_09_2025_21_54
2025-09-20 21:54:59,600 - INFO - Session directory created: /home/monkey/razvand/orgs/unikraft/autokraft/.sessions/catalog/examples/httpserver-nodejs21/session_20_09_2025_21_54
2025-09-20 21:54:59,600 - INFO - Session directory created: /home/monkey/razvand/orgs/unikraft/autokraft/.sessions/catalog/examples/httpserver-nodejs21/session_20_09_2025_21_54
2025-09-20 21:54:59,600 - INFO - Session initialized: session_20_09_2025_21_54
2025-09-20 21:54:59,600 - INFO - Session initialized: session_20_09_2025_21_54
2025-09-20 21:54:59,600 - INFO - Using tests directory: .tests
2025-09-20 21:54:59,600 - INFO - Using tests directory: .tests
2025-09-20 21:54:59,600 - INFO - Using app directory: .app
2025-09-20 21:54:59,600 - INFO - Using app directory: .app
2025-09-20 21:54:59,811 - INFO - Generating runtimes...node/21:local
2025-09-20 21:54:59,811 - INFO - Generating runtimes...node/21:local
2025-09-20 21:54:59,811 - INFO - Catalog runtime path: /home/monkey/razvand/orgs/unikraft/catalog/library/node/21
2025-09-20 21:54:59,811 - INFO - Catalog runtime path: /home/monkey/razvand/orgs/unikraft/catalog/library/node/21
2025-09-20 21:55:02,718 - INFO - Generated 0 target configuration(s) successfully.
2025-09-20 21:55:02,718 - INFO - Generated 0 target configuration(s) successfully.
2025-09-20 21:55:02,718 - INFO - Generating key target runtimes for example application.
2025-09-20 21:55:02,718 - INFO - Generating key target runtimes for example application.
2025-09-20 21:55:02,731 - INFO - Running app_init_fs.sh for example application.
2025-09-20 21:55:02,731 - INFO - Running app_init_fs.sh for example application.
2025-09-20 21:55:02,731 - INFO - Running app_fs_init.sh
2025-09-20 21:55:02,731 - INFO - Running app_fs_init.sh
2025-09-20 21:55:03,054 - INFO - Initialized application filesystem initrd.cpio is stored in /home/monkey/razvand/orgs/unikraft/autokraft/initrd.cpio
2025-09-20 21:55:03,054 - INFO - Initialized application filesystem initrd.cpio is stored in /home/monkey/razvand/orgs/unikraft/autokraft/initrd.cpio
2025-09-20 21:55:03,054 - INFO - Copying log file app_fs_init.log to session directory.
2025-09-20 21:55:03,054 - INFO - Copying log file app_fs_init.log to session directory.
2025-09-20 21:55:03,055 - INFO - Completed 0 test(s) successfully.
2025-09-20 21:55:03,055 - INFO - Completed 0 test(s) successfully.
```

razvand commented:
Make sure all files and with a newline (motivation [here](https://stackoverflow.com/a/729795/4804196)).

You can see those files not ending in a newline by browsing the [`Files changed` screen of the PR](https://github.com/unikraft/autokraft/pull/1/files). They appear as the attached image.

<img width="369" height="103" alt="2025-09-20-230429_369x103_scrot" src="https://github.com/user-attachments/assets/ad144bb0-31f7-4966-97fe-80d2bb547627" />

razvand commented:
Additionally, be sure you don't have extra blank lines the end of the file, nor do you have trailing whitespaces on files.

shank250 committed:
fix: Update curl run command for bridge & tap

Update the port number for curl command with bridge & tap networking to use
ExposedPort (default 80) from RunConfig.yaml.

Sign-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
docs: Update README.md with latest repository link

Update Set up instruction in README.md with latest repository link.

Sign-off-by: Shashank Srivastava <shashank21005@gmail.com>

shank250 committed:
fix: Update all the remove commands

Update remove commands with prefix sudo to remove builds from
previous phase with root access.

Sign-off-by: Shashank Srivastava <shashank21005@gmail.com>

razvand reviewed (approved):
Amazing work, @shank250 :partying_face: 

Let's make us of `autokraft` as much as possible from now on.
